### ISSUE TITLE
Invalid type definition on subprocess.run with partial

### ISSUE THREAD

**Issue by lambdalisue** [microsoft/pyright#455]
**Describe the bug**
It seems `subprocess.run`'s return type become `CompleteProcess[str]` forcedly when used with `partial`

**To Reproduce**
Steps to reproduce the behavior.

```
$ pyright test.py                                                                                                                                                                                                                                                               
Searching for source files
Found 1 source file
/Users/alisue/test.py
  27:20 - error: Cannot access member 'decode' for type 'str'
  Member 'decode' is unknown
1 error, 0 warnings
Completed in 1.257sec

$ mypy test.py                                                                                                                                                                                                                                                            
Success: no issues found in 1 source file
```
**Expected behavior**

The type of `r` in `incorrect` function become `CompleteProcess[bytes]` so `r.stdout.decode` works.

**Screenshots or Code**

```python
from __future__ import annotations

import subprocess
from functools import partial


def correct1():
    r = subprocess.run(["echo", "Hello"], stdout=subprocess.PIPE)
    print(r.stdout.decode("utf-8"))


def correct2():
    r = subprocess.run(["echo", "Hello"], stdout=subprocess.PIPE, encoding="utf-8")
    print(r.stdout.encode("utf-8").decode("utf-8"))


def correct3():
    r = partial(
        subprocess.run, ["echo", "Hello"], stdout=subprocess.PIPE, encoding="utf-8"
    )()
    print(r.stdout.encode("utf-8").decode("utf-8"))


def incorrect():
    r = partial(subprocess.run, ["echo", "Hello"], stdout=subprocess.PIPE)()
    print(r.stdout.decode("utf-8"))


correct1()
correct2()
correct3()
incorrect()
```

**VS Code extension or command-line**
command-line

**Additional context**

**Comment by erictraut** [comment]
I think Pyright is doing the correct thing here. The problem is that type information is lost when you call `partial`. The `run` method contains multiple overloaded definitions in `subprocess.pyi`. If you call `run` directly (as opposed to passing it as an argument to `partial`), a type checker will pick the first overloaded definition that matches the provided arguments. Because you're passing `subprocess.run` to the `partial` function, it is matching against the parameter `Callable[..., _T]` in the `partial` function, and it's picking the first overload that matches.

You can see that if you call `subprocess.run` directly, it works as you expect:
```
r = subprocess.run(["echo", "Hello"], stdout=subprocess.PIPE).stdout.decode('utf-8')
```

If you want to make use of strong type checking, I encourage you to avoid using the `functools.partial`. You can do so by creating your own properly-typed function that calls subprocess.run with prespecified arguments. This function can be defined in a closure if those argument values need to be dynamically calculated.

Another (less elegant) workaround is to call `cast` to override the type inferred by the type checker.

**Comment by calops** [comment]
What about this issue? Seems similar, but I can't imagine a workaround beyond specifying every single parameter in the function signature (but even then I find it weird that this issue happens in the first place):

```python
import subprocess


def call():
    return subprocess.run('ls')


def call_with_args(**kwargs):
    return subprocess.run('ls', **kwargs)


# This works properly
call().stdout.decode()

# This does not, with the same typing error as described in the initial issue
call_with_args(stdout=subprocess.PIPE).stdout.decode()
```

For some reason, as soon as `kwargs` are involved, pyright detects the return type of `run()` as `CompletedProcess[str]` instead of `CompletedProcess[bytes]` otherwise.

Should this be reported in a separate issue, or can we reopen this one?

**Comment by erictraut** [comment]
>For some reason, as soon as kwargs are involved, pyright detects the return type of run() as CompletedProcess[str] instead of CompletedProcess[bytes] otherwise.

That's because of the way `overload` works. When a type checker encounters a call expression and that call is an overloaded function, it evaluates each overload in the order they are declared. If the arguments don't match the overload, it moves to the next one, etc. If the number and type of the arguments is completely unknown (as they are with `**kwargs`), a type checker will assume that the first overload is applicable. The first overload for `subprocess.run` returns a `CompletedProcess[str]`. 

In effect, you're "blinding" the type checker when you use `**kwargs` because there's no type information it can use to resolve the overload.

Pyright is doing the right thing here, so there's no need to file a new bug report.
